### PR TITLE
Intake

### PR DESCRIPTION
**Pull request by SaltyJoke**
Includes hopper because hopper and intake should run simultaneously

### CODE REVIEW

**Comment by SaltyJoke** [comment]
Jack says no functional problems with the structure, but odd command names. Should be fine.

File: src/main/java/org/team199/robot2020/commands/ChangedDesiredIntake.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+package org.team199.robot2020.commands;
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+
+// NOTE:  Consider using this command inline, rather than writing a subclass.  For more
+// information, see:
+// https://docs.wpilib.org/en/latest/docs/software/commandbased/convenience-features.html
+public class ChangedDesiredIntake extends InstantCommand {
```
Comment (kevinzwang) [inline src/main/java/org/team199/robot2020/commands/ChangedDesiredIntake.java]: I have no idea what this class does based on the name. What is the desired intake and why are we changing it? Add comments or make the name more clear please

File: src/main/java/org/team199/robot2020/commands/ChangedDesiredIntake.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  double buttonPressed;
+
+  public ChangedDesiredIntake(double buttonPressed) {
+    this.buttonPressed=buttonPressed;
+  }
+
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize() {
+    if (buttonPressed == SmartDashboard.getNumber("Desired Intake Deployment", 0.0 )) {
```
Comment (kevinzwang) [inline src/main/java/org/team199/robot2020/commands/ChangedDesiredIntake.java]: the SmartDashboard value "Desired Intake Deployment" is a boolean in RobotContainer. Be consistent with this stuff. In general I wouldn't even recommend using SmartDashboard to store numbers

File: src/main/java/org/team199/robot2020/commands/UpdateIntake.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+/*----------------------------------------------------------------------------*/
+
+package org.team199.robot2020.commands;
+
+import org.team199.robot2020.subsystems.Intake;
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+public class UpdateIntake extends CommandBase {
```
Comment (kevinzwang) [inline src/main/java/org/team199/robot2020/commands/UpdateIntake.java]: I cannot understand at all what this command does

File: src/main/java/org/team199/robot2020/commands/RunIntakeAndHopper.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package org.team199.robot2020.commands;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import org.team199.robot2020.subsystems.Intake;
+
+public class RunIntakeAndHopper extends CommandBase {
```
Comment (kevinzwang) [inline src/main/java/org/team199/robot2020/commands/RunIntakeAndHopper.java]: Change Hopper to Feeder

**Review by kevinzwang — CHANGES_REQUESTED** [review]
needs work before final code is written

**Comment by kevinzwang** [comment]
Oof I may have screwed up something small when merging dev into this branch. Will try to fix later

**Comment by SaltyJoke** [comment]
I suggest replacing the SmartDashboard and all the variables with an enum (see last year's climber). Also why does the intake go up when outtaking?

**Comment by SaltyJoke** [comment]
What buttons will be used? Will there be sensors to detect/count balls? Should ChangeDesiredIntake and UpdateIntake be the same class? What happened to DeployIntake and RunIntakeAndFeeder's function?

File: src/main/java/org/team199/robot2020/subsystems/Intake.java [hunk 5]
```diff
@@ -11,29 +11,34 @@
...
+
+    public void deploy() {
+        intakePistons.set(DoubleSolenoid.Value.kForward);
+    }
+
+    public void retract() {
+        intakePistons.set(DoubleSolenoid.Value.kReverse);
+    }
+
+    public boolean isRunning() {
```
Comment (SaltyJoke) [inline src/main/java/org/team199/robot2020/subsystems/Intake.java]: Why is this method's functionality and the way it's used the opposite of what the name suggests?

File: src/main/java/org/team199/robot2020/commands/Regurgitate.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  @Override
+  public void execute() {
+  }
+
+  // Called once the command ends or is interrupted.
+  @Override
+  public void end(boolean interrupted) {
+    feeder.stop();
+    if (intake.isDeployed()) {
+      intake.intake();
```
Comment (SaltyJoke) [inline src/main/java/org/team199/robot2020/commands/Regurgitate.java:46]: Aren't there any situations where you would want to regurgitate without starting to intake again?

File: src/main/java/org/team199/robot2020/subsystems/Feeder.java [hunk 7]
```diff
@@ -16,34 +16,40 @@
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 
 public class Feeder extends SubsystemBase {
-  private final WPI_TalonSRX beltMotor = MotorControllerFactory.createTalon(Constants.Feeder.kBeltMotor);
-  private final WPI_TalonSRX ejectMotor = MotorControllerFactory.createTalon(Constants.Feeder.kEjectMotor);
-  private final TimeOfFlight indexSensor = new TimeOfFlight(Constants.Feeder.kIndexSensor);
+  private static final double kBeltIntakeSpeed = .8;
+  private static final double kBeltEjectSpeed = 1;
+  private static final double kRollerEjectSpeed = 1;
+  private static final double kIndexerDistance = 127; // 5 inches in millimeters
```
Comment (SaltyJoke) [inline src/main/java/org/team199/robot2020/subsystems/Feeder.java]: Perhaps make it inches times the factor so it can be changed easily?

**Review by SaltyJoke — COMMENTED** [review]
I think it's good enough to merge.

File: src/main/java/org/team199/robot2020/commands/Regurgitate.java (hunk 6 above)
Comment (kevinzwang) [reply src/main/java/org/team199/robot2020/commands/Regurgitate.java:46]: wdym?

File: src/main/java/org/team199/robot2020/subsystems/Feeder.java (hunk 7 above)
Comment (kevinzwang) [reply src/main/java/org/team199/robot2020/subsystems/Feeder.java]: sounds good

**Review by kevinzwang — DISMISSED** [review]

**Pull request merged**